### ISSUE TITLE
False positive type error when passing `AnyStr` value to `re.search`

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#5856]
### Discussed in https://github.com/microsoft/pyright/discussions/5850

<div type='discussions-op-text'>

<sup>Originally posted by **Evan6998** August 30, 2023</sup>
I wonder if I can annotate a function parameters with generic `TypeVar` and pass them to `re.search`?
What I wanna do is as simple as : 
```python
import re
from typing import TypeVar

T = TypeVar("T", str, bytes)

def foo(a: T, b: T):
    ...
    re.search(a, b)
    ...
```

However, `re.search` seems to regard T as sum types.
<img width="1075" alt="image" src="https://github.com/microsoft/pyright/assets/17891367/2967ef6d-cf30-4096-83d8-5ef1fcc314cc">

I notice that `overload` might work but I wonder whether generic can do the same ?</div>

**Comment by erictraut** [comment]
Pyright doesn't support this particular use of a constrained type variable. The only way I can think to make this work would have significant performance impact on the evaluation of all call expressions, regardless of whether they involve constrained type variables. 

In [this documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=constrained-type-variables), I say "This approach allows pyright to handle some constrained TypeVar use cases that mypy cannot, but there are conversely other use cases that mypy can handle and pyright cannot." This is an example of a case that it cannot handle.

You'll need to find a workaround in your code if you want this to type check with pyright.
